File: python/packages/microsoft/src/matimo_microsoft/tools/ms_get_email/ms_get_email.py
"""
ms_get_email — GET /me/messages
https://learn.microsoft.com/en-us/graph/api/user-list-messages
Mirrors: typescript/packages/microsoft/tools/ms_get_email/ms_get_email.ts
"""
from __future__ import annotations

import math
from typing import Any
from urllib.parse import quote

from matimo.errors import ErrorCode, MatimoError
from matimo_microsoft.graph_client import get_access_token, graph_request, require_params

_DEFAULT_TOP = 10
_MAX_TOP = 50


def _format_sender(message: dict[str, Any]) -> str:
    address = (message.get("from") or {}).get("emailAddress")
    if not address:
        return ""
    name = address.get("name")
    email = address.get("address")
    if name and email:
        return f"{name} <{email}>"
    return name or email or ""


async def run(params: dict[str, Any]) -> dict[str, Any]:
    require_params(params, [], "ms_get_email")

    raw_top = params.get("top")
    top = _DEFAULT_TOP if raw_top is None else raw_top
    try:
        top_number = float(top)
    except (TypeError, ValueError):
        top_number = math.nan
    if not math.isfinite(top_number) or top_number < 1 or top_number > _MAX_TOP:
        raise MatimoError(
            f"ms_get_email: 'top' must be a number between 1 and {_MAX_TOP} (received {top!r})",
            ErrorCode.VALIDATION_FAILED,
            {"top": raw_top},
        )

    folder_id_param = params.get("folder_id")
    folder_id = folder_id_param if isinstance(folder_id_param, str) and folder_id_param else None
    filter_param = params.get("filter")
    filter_expr = filter_param if isinstance(filter_param, str) and filter_param else None
    search_param = params.get("search")
    search_expr = search_param if isinstance(search_param, str) and search_param else None

    token = get_access_token(params)

    path = (
        f"/me/mailFolders/{quote(folder_id, safe='')}/messages" if folder_id else "/me/messages"
    )

    query: dict[str, Any] = {
        "$top": int(top_number),
        "$select": "id,subject,from,receivedDateTime,isRead,bodyPreview,hasAttachments",
    }
    if filter_expr:
        query["$filter"] = filter_expr
    if search_expr:
        query["$search"] = search_expr

    data = await graph_request(
        method="GET",
        path=path,
        token=token,
        resource_type="Mail folder",
        query=query,
        headers={"ConsistencyLevel": "eventual"} if search_expr else None,
    )
    raw_messages = data.get("value") if isinstance(data, dict) else None

    messages = [
        {
            "id": message.get("id") or "",
            "subject": message.get("subject") or "",
            "from": _format_sender(message),
            "received_at": message.get("receivedDateTime") or "",
            "is_read": message.get("isRead") or False,
            "body_preview": message.get("bodyPreview") or "",
            "has_attachments": message.get("hasAttachments") or False,
        }
        for message in raw_messages or []
    ]

    return {
        "success": True,
        "messages": messages,
    }

File: typescript/packages/microsoft/tools/ms_get_email/ms_get_email.ts
/**
 * ms_get_email — GET /me/messages
 * https://learn.microsoft.com/en-us/graph/api/user-list-messages
 */
import { MatimoError, ErrorCode } from '@matimo/core';
import { getAccessToken, requireParams, graphRequest, type ToolContext } from '../graph-client';

const DEFAULT_TOP = 10;
const MAX_TOP = 50;

interface EmailAddress {
  name?: string;
  address?: string;
}

interface MessageRecipient {
  emailAddress?: EmailAddress;
}

interface GraphMessage {
  id?: string;
  subject?: string;
  from?: MessageRecipient;
  receivedDateTime?: string;
  isRead?: boolean;
  bodyPreview?: string;
  hasAttachments?: boolean;
}

interface MessagesResponse {
  value?: GraphMessage[];
}

function formatSender(message: GraphMessage): string {
  const address = message.from?.emailAddress;
  if (!address) return '';
  if (address.name && address.address) return `${address.name} <${address.address}>`;
  return address.name ?? address.address ?? '';
}

export default async function execute(
  params: Record<string, unknown>,
  context?: ToolContext
): Promise<unknown> {
  requireParams(params, [], 'ms_get_email');

  const top = params.top === undefined ? DEFAULT_TOP : Number(params.top);
  if (!Number.isFinite(top) || top < 1 || top > MAX_TOP) {
    throw new MatimoError(
      `ms_get_email: 'top' must be a number between 1 and ${MAX_TOP} (received ${String(params.top)})`,
      ErrorCode.VALIDATION_FAILED,
      { top: params.top }
    );
  }

  const folderId = typeof params.folder_id === 'string' && params.folder_id ? params.folder_id : undefined;
  const filter = typeof params.filter === 'string' && params.filter ? params.filter : undefined;
  const search = typeof params.search === 'string' && params.search ? params.search : undefined;

  const token = getAccessToken(context);

  const path = folderId
    ? `/me/mailFolders/${encodeURIComponent(folderId)}/messages`
    : '/me/messages';

  const data = await graphRequest<MessagesResponse>({
    method: 'GET',
    path,
    token,
    resourceType: 'Mail folder',
    query: {
      $top: top,
      $select: 'id,subject,from,receivedDateTime,isRead,bodyPreview,hasAttachments',
      ...(filter ? { $filter: filter } : {}),
      ...(search ? { $search: search } : {}),
    },
  });

  const messages = (data?.value ?? []).map((message) => ({
    id: message.id ?? '',
    subject: message.subject ?? '',
    from: formatSender(message),
    received_at: message.receivedDateTime ?? '',
    is_read: message.isRead ?? false,
    body_preview: message.bodyPreview ?? '',
    has_attachments: message.hasAttachments ?? false,
  }));

  return {
    success: true,
    messages,
  };
}
